components/session-service/server/server.go

```diff
@@ -347,15 +352,89 @@ func (s *Server) logoutHandler(w http.ResponseWriter, r *http.Request) {
 	w.WriteHeader(http.StatusOK)
 }
 
+func filterUserPolicy(policyId string, policyIds []string) bool {
+	for _, _policyId := range policyIds {
+		if _policyId == policyId {
+			return true
+		}
+	}
+	return false
+}
+
 func (s *Server) userPoliciesHandler(w http.ResponseWriter, r *http.Request) {
 	decoder := json.NewDecoder(r.Body)
 	var data userData
 	err := decoder.Decode(&data)
 	if err != nil {
-		fmt.Println(err, "error")
+		http.Error(w, errors.Wrap(err, "failed to decode").Error(),
+			http.StatusInternalServerError)
+		return
+	}
+
+	teamsForUser, err := s.teamsServiceClient.GetTeamsForMember(r.Context(), &teams.GetTeamsForMemberReq{
+		UserId: data.UserId,
+	})
+	if err != nil {
+		http.Error(w, errors.Wrap(err, "failed teamsServiceClient").Error(),
+			http.StatusInternalServerError)
+		return
+	}
+
+	var teamsPolicyIds []string
+	if len(teamsForUser.Teams) > 0 {
+		for _, team := range teamsForUser.Teams {
+			teamPolicies, err := s.authzPoliciesClient.GetUserPolicies(r.Context(), &authz.GetUserPoliciesReq{
+				MemberType:  "team",
+				Username:    team.Id,
+				ConnectorId: data.ConnectorID,
+			})
+			if err != nil {
+				http.Error(w, errors.Wrap(err, "failed to GetUserPolicies").Error(),
+					http.StatusInternalServerError)
+				return
+			}
+			teamsPolicyIds = append(teamsPolicyIds, teamPolicies.PolicyIds...)
+		}
 	}
 
-	fmt.Println(data, "dataAZ_")
+	userPolicies, err := s.authzPoliciesClient.GetUserPolicies(r.Context(), &authz.GetUserPoliciesReq{
+		MemberType:  "user",
+		Username:    data.Username,
+		ConnectorId: data.ConnectorID,
+	})
+	if err != nil {
+		http.Error(w, errors.Wrap(err, "failed to GetUserPolicies").Error(),
+			http.StatusInternalServerError)
+		return
+	}
+
+	combinedPolicyIds := append(userPolicies.PolicyIds, teamsPolicyIds...)
+
+	listOfPolicies, err := s.authzPoliciesClient.ListPolicies(r.Context(), &authz.ListPoliciesReq{})
+	if err != nil {
+		http.Error(w, errors.Wrap(err, "failed to get listOfPolicies").Error(),
+			http.StatusInternalServerError)
+		return
+	}
+
+	var _projectRolePairs []projectRolePairs
+	for _, policy := range listOfPolicies.Policies {
```

Comment:
this has a time complexity of O(n^2), which can be reduced to constant time using maps. Please store listOfPolicies in a map and find the combinedPolicyId in that map.

Reply:
got it.